### PR TITLE
Use cython to compile jsonobject

### PR DESCRIPTION
**Pull request by emord**
@czue @esoergel catching up on some reading/articles and saw an idea I wanted to try out. This is in response to discussion here: https://gist.github.com/czue/6580a94b37e18d3e587ded5767922b96

This requires no code changes and I bet there's a bigger gain to be had if we look through cython docs and start declaring types. In a very unscientific test I noticed the jsonobject tests running in half the time, and corehq.apps.userreports tests having a ~15% speedup. HQ PR here: https://github.com/dimagi/commcare-hq/pull/18955

### CODE REVIEW

**Comment by esoergel** [comment]
Awesome!  I'm not sure how to review this exactly, but it sounds reasonable.  I bet @dannyroberts will want to weigh in.

**Comment by dannyroberts** [comment]
Wow, what a sweet thing to find in my inbox!

So, what I'm seeing is a couple changes to the setup.py some harmless extra things to .gitignore, and then a bunch of what looks like auto-generated cython files.

Ideally what I'd _like_ to see is something explaining what series of commands you need to run to get those c files to be generated, and then _not_ check them in (if possible) and have that get run by `setup.py sdist upload` or whatever we use nowadays to push to pip. Or, if checking in these C / pyx files is a must, having a travis test that runs the build step and at least asserts that the checked-in files match the freshly-built files.

Anyway, this is totally cool! Years ago when I first wrote this I think I tried this and whatever I was using (cython? pypy?) wasn't quite there yet—or at least I couldn't get it working. Can't wait to get everything using this!

**Comment by emord** [comment]
> you need to run to get those c files to be generated, and then not check them in (if possible) 

I used cythonize as mentioned here: http://cython.readthedocs.io/en/latest/src/reference/compilation.html You're supposed to check the C files in so that people don't need to install cython if they don't want to

with that said I htink I may have only half-done the correct thing here: http://cython.readthedocs.io/en/latest/src/reference/compilation.html#distributing-cython-modules

I'd need to do a more thorough look through those docs to make sure this is doing the correct thing. I can pick this up in January when I'm back unless one of you three gets inspired to take this up.

**Comment by czue** [comment]
this is great. :fire: 

> I can pick this up in January when I'm back unless one of you three gets inspired to take this up.

I've got plenty to keep myself busy so will let others get this over the hump as I don't know much / anything about best practices in this space.

**Comment by dannyroberts** [comment]
Yeah, that's awesome. Obviously no urgency here—that was just me getting excited. Looking forward to helping roll this out after the holidays! 🎄

**Comment by emord** [comment]
> Ideally what I'd like to see is something explaining what series of commands you need to run to get those c files to be generated, 

https://github.com/dimagi/jsonobject/pull/141/commits/643947e1b455722ce741ec54da505c5c1685b808

> and then not check them in (if possible) and have that get run by setup.py sdist upload or whatever we use nowadays to push to pip. Or, if checking in these C / pyx files is a must, 

as mentioned earlier it's good to check in the C files so that it doesn't depend on Cython to be installed in the project using this

> having a travis test that runs the build step and at least asserts that the checked-in files match the freshly-built files.

I think this could be a script that does:

```bash
find jsonobject -iname '*.c' -delete
python setup.py build_ext --inplace
git diff doesn't have anything
```

Does that sound like a sufficient test?

File: setup.py [hunk 1]
```diff
@@ -1,19 +1,45 @@
 from setuptools import setup
 import io
 
+from setuptools.extension import Extension
+
+try:
+    # Only use Cython it's installed in the environment, otherwise use the provided C
```
Comment (dannyroberts) [inline setup.py]: nit: ^ _if_ it's.

**Comment by emord** [comment]
Added a test. I think it is deterministic and will only change with Cython versions or code changes. Testing this out on staging for a couple of days before pushing it out

**Comment by czue** [comment]
I've not done any profiling recently but am happy to do that. Just thinking out loud - is there any chance that this is not a good idea assuming it's bug-free? You've already profiled it at a low level to validate that it's definitely faster, right? Just wondering if we could just test and merge so as to basically just run a pre/post test that acts as a profile.

**Comment by emord** [comment]
Been running into some issues deploying on staging that may be related to this change. Going to look into it in the next couple of days

**Comment by emord** [comment]
@dannyroberts I removed the test that tested that *.c files were up to date because it was failing on python3. I made a python 3 virtualenv locally and saw that there are some small line ending differences between running cythonize on python 2 vs 3, so py2 tests were passing, but py3 weren't. Example build: https://travis-ci.org/dimagi/jsonobject/builds/329228989?utm_source=github_status&utm_medium=notification

I have the branch with a test here: https://github.com/dimagi/jsonobject/tree/je/old-cythonize-with-test but this seems like a small thing that shouldn't block release as this library is pretty stable. Do you think this is a blocker for merging and releasing?

**Comment by dannyroberts** [comment]
Hmm, could you have it only run for python 2.7 (i.e. skip for python 3)? I'm not sure if travis gives you access to that in an env var or something, but if it doesn't there's always `python --version` which prints "Python {major}.{minor}.{bugfix}" to stderr. So you could just replace

```
./test_cython_files.sh
```

with

```
- |
  if [ "$(python --version 2>&1 | cut -d'.' -f1,2)" = "Python 2.7" ]
  then
    ./test_cython_files.sh
  else
    echo "Skipping cython build test on python 3"
  fi
```

If that ends up being its own rabbithole, I suppose it's _okay_ to push this out without that test, but I'm also wary of doing that, since in my opinion the error that test is supposed to guarantee against is almost guaranteed to happen without it. If the above solution works, I think it's worth including in this PR.

**Comment by emord** [comment]
Thanks for the example code. Tests are passing. going to merge. will release to HQ tomorrow

**Comment by dannyroberts** [comment]
🎉 excited about this!

**Comment by czue** [comment]
awesome! :clap:

**Comment by millerdev** [comment]
This looks great! I've suspected for a long time that JsonObject is slowing us down, but never made time to actually look into it.

**Comment by emord** [comment]
@millerdev FWIW I think there are still some good gains to be had there. At some point we should sit down and go through cython docs to see what we can do. From a talk from Instagram really all you need to do is start using cdef/cpdef and begin declaring types to get most/all the gains you need

**Comment by emord** [comment]
First look at a particular metric after this PR @czue @dannyroberts @esoergel @millerdev 

cython_jsonobject

**Comment by emord** [comment]
Smoothed graph:

smoothed_cython

went from ~85s to ~65s for about 20% gain on ICDS UCRs

**Comment by czue** [comment]
this is awesome! :raised_hands:

**Comment by dannyroberts** [comment]
Wow, at the risk of engaging in the hyperbole of the century: that is not nothing! 📉 🎉

**Pull request merged**